lib/ex_aws/ecs.ex

```diff
@@ -1021,16 +1049,48 @@ defmodule ExAws.ECS do
 
   ## Private parts
 
-  # Pascalize has room for improvements.. but it seems to work
+  defp transform_list(opts, list_key, value_key, fn_transform) do
+    transformed = opts
+    |> Keyword.get(list_key, [])
+    |> transform_list(value_key, fn_transform)
 
-  defp pascalize_keys(opts) when is_map(opts) do
+    Keyword.put(opts, list_key, transformed)
+  end
+  defp transform_list(list, value_key, fn_transform) do
+    list
+    |> Enum.map(&(transform_value(&1, value_key, fn_transform)))
+  end
+
+  defp transform_value(map, key, fn_transform) when is_map(map) do
+    map
+    |> Keyword.new
+    |> transform_value(key, fn_transform)
+    |> Map.new
+  end
+  defp transform_value(opts, key, fn_transform) do
     opts
-    |> Enum.map(&pascalize_keys/1)
-    |> Enum.into(%{})
+    |> Keyword.get_and_update(key, fn(value) ->
+        new_val = fn_transform.(value)
+        {value, new_val}
+      end)
+    |> update_only_if_existed(opts)
+  end
+
+  defp update_only_if_existed({nil, _}, original), do: original
+  defp update_only_if_existed({_, transformed}, _), do: transformed
+
+  defp replace_underscore_with_dash(value) when is_atom(value) do
+    value
+    |> Atom.to_string
+    |> replace_underscore_with_dash
+  end
+  defp replace_underscore_with_dash(value) when is_binary(value), do: String.replace(value, "_", "-")
+
+  defp pascalize_keys(opts) when is_map(opts) do
+    Map.new(opts, &pascalize_keys/1)
   end
-  defp pascalize_keys([]), do: []
-  defp pascalize_keys([head | tail]) do
-    [pascalize_keys(head)] ++ pascalize_keys(tail)
+  defp pascalize_keys(list) when is_list(list) do
+    list |> Enum.map(&pascalize_keys/1)
```

Comment:
It's recursive and options can be deeply nested, with both maps and lists of maps etc. For an example of complex options, see [this test](https://github.com/larskrantz/ex_aws/blob/08fbc80655ac266d21e7090c2e8d86b23beb2622/test/lib/ex_aws/ecs_test.exs#L306-L347). We need to traverse the options and handle all of the keys and values deep down.

This is how I think: It starts with a map. The map has keys and values. Keys are pascalized using `pascalize_key/1` and values are recursively passed on to `pascalize_keys/1`. If it's a list, traverse and handle values. If a value is another list or map, recurse. If the value is a plain value, transform it if it's needed (atom to string) or just return value in the catch-all clause.
I can't really see how to use different function names for the list and values when recursing.

I'm sorry if it's confusing, but I'm totally open to suggestions as long as tests pass.